### ISSUE TITLE
support Ephemeral models

### ISSUE THREAD

**Issue by dataders** [dbt-msft/dbt-sqlserver#166]
from #137, here's the proposed solution

> ### Ephemeral models
> 
> The _general_ approach outlined in https://github.com/dbt-msft/dbt-synapse/commit/9180da5a6d092c6b4b22d5388aa6592d2b9d77c4 is still the same, though I was doing it on easy mode, and didn't actually try to solve for ephemeral model compilation :)
> 
> The mechanism is: [reimplement the compiler](https://github.com/dbt-msft/dbt-synapse/commit/9180da5a6d092c6b4b22d5388aa6592d2b9d77c4#diff-f5edefcace852d4b7f256956b00c049883a49be1d8216dd2438480b8f1e55c09R43-R45), define a new one (like in `dbt/adapters/sqlserver/compilation.py`), and reimplement methods like [`_inject_ctes_into_sql `](https://github.com/fishtown-analytics/dbt/blob/abe8e839458d0288f76cb6950e79c77e3b1627cc/core/dbt/compilation.py#L191`) and [`_recursively_prepend_ctes`](https://github.com/fishtown-analytics/dbt/blob/abe8e839458d0288f76cb6950e79c77e3b1627cc/core/dbt/compilation.py#L252).
> 
> Taking a step back: In order to support ephemeral models with CTEs on SQLServer and Synapse, given that T-SQL doesn't support CTEs nested inside either CTEs or subqueries, would we have to take the same approach as the one outlined above?
> 
> Namely, instead of recursively prepending ephemeral models as CTEs to the start of the query:
> ```sql
> with  __dbt__cte__ephemeral_model as (
> 
> 
> with my_cte as (
>     
>     select 1 as id
>     
> )
> 
> select * from my_cte
> ),ephemeral_model as (
> 
> select * from __dbt__cte__ephemeral_model
> 
> ),
> 
> another_cte as (
>     
>     select 2 as id
>     
> )
> 
> select * from ephemeral_model
> union all
> select * from another_cte
> ```
> 
> Recursively prepend them as temp tables, to be executed in-transaction with the body of the query:
> ```sql
> create table #ephemeral_model as (
> 
> 
> with my_cte as (
>     
>     select 1 as id
>     
> )
> 
> select * from my_cte
> );
> 
> with ephemeral_model as (
> 
> select * from #ephemeral_model
> 
> ),
> 
> another_cte as (
>     
>     select 2 as id
>     
> )
> 
> select * from ephemeral_model
> union all
> select * from another_cte
> ```

**Comment by jtcohen6** [comment]
I don't think we should do this :)

Newer MSFT databases should support nested CTEs (https://github.com/dbt-msft/dbt-sqlserver/issues/457). We shouldn't expose the full `Compiler` interface to adapters (https://github.com/dbt-labs/dbt-core/pull/9134).

**Comment by joshmarkovic** [comment]
@dataders, is this still something you want today? What exactly is your use case? I don't like the compiler-override because it's discouraged by core and semantically wrong...

**Comment by joshmarkovic** [comment]
@axellpadilla, do you thoughts on whether we want to keep this open?

**Comment by axellpadilla** [comment]
@joshmarkovic Yes, this is an interesting feature, a good implementation (that doesn't cause blocking issues or re implementing the whole compiler), will leave us closer to official docs https://docs.getdbt.com/docs/build/materializations?version=2 , that said, looks like official docs is just a cte and no disk io
